src/runner.rs

```diff
@@ -46,79 +90,147 @@ pub async fn run_test_command(
+                Err(e) => {
+                    log::warn!("Error opening file: {}", e);
+                    continue;
+                }
+            };
+            let reader = std::io::BufReader::new(file);
+            let junitxml = match junit_parser::from_reader(reader) {
+                Ok(junitxml) => junitxml,
+                Err(e) => {
+                    log::warn!("Error parsing junitxml: {}", e);
+                    continue;
+                }
+            };
+            for suite in junitxml.suites {
+                let parent_name = suite.name;
+                for case in suite.cases {
+                    let failure = case.status.is_failure();
+                    if failure {
+                        let name = case.original_name;
+                        log::debug!("Test failed: {} -> {}", parent_name, name);
+                        failures.push(Test {
+                            parent_name: parent_name.clone(),
+                            name: name.clone(),
+                            class_name: case.classname.clone(),
+                            file: case.file.clone(),
+                        });
                     }
                 }
             }
         }
     }
-    let exit_code = result.code().unwrap_or(EXIT_FAILURE);
-    Ok(RunResult {
+    Ok(failures)
+}
+
+pub async fn run_quarantine(
+    run_result: &RunResult,
+    api_address: &str,
+    token: &str,
+    org_url_slug: &str,
+    repo: &BundleRepo,
+    delay: std::iter::Take<ExponentialBackoff>,
+    no_quarantining: &bool,
+) -> anyhow::Result<QuarantineRunResult> {
+    if *no_quarantining {
+        log::info!("Skipping quarantining step.");
+        return Ok(QuarantineRunResult {
+            exit_code: run_result.exit_code,
+            quarantine_status: QuarantineBulkTestStatus {
+                group_is_quarantined: false,
+                quarantine_results: Vec::new(),
+            },
+        });
+    }
+    // check with the API if the group is quarantined
+    log::info!(
+        "Checking quarantine status for failures: {:?}",
+        run_result.failures
+    );
+    let quarantine_results = if run_result.failures.is_empty() {
+        QuarantineBulkTestStatus {
+            group_is_quarantined: false,
+            quarantine_results: Vec::new(),
+        }
+    } else {
+        match Retry::spawn(delay, || {
+            get_quarantine_bulk_test_status(
+                api_address,
+                token,
+                org_url_slug,
+                &repo.repo,
+                &run_result.failures,
+            )
+        })
+        .await
+        {
+            Ok(quarantine_results) => quarantine_results,
+            Err(e) => {
+                log::error!("Failed to get quarantine results: {:?}", e);
+                QuarantineBulkTestStatus {
+                    group_is_quarantined: false,
+                    quarantine_results: Vec::new(),
+                }
+            }
+        }
```

Comment:
similar to the above, this is a very verbose way of handling errors. Rather, we can use `unwrap_or_else` more idiomatically:

```suggestion
        .await
        .unwrap_or_else(|e| {
            log::error!("Failed to get quarantine results: {:?}", e);
            QuarantineBulkTestStatus {
                group_is_quarantined: false,
                quarantine_results: Vec::new(),
            }
        })
```